lib/IRGen/LocalTypeData.cpp

```diff
@@ -381,10 +382,103 @@ IRGenFunction::setScopedLocalTypeMetadataForLayout(SILType type,
   setScopedLocalTypeData(key, response);
 }
 
-void IRGenFunction::setScopedLocalTypeMetadata(CanType type,
-                                               MetadataResponse response) {
+namespace {
+
+void setScopedLocalTypeMetadataImpl(IRGenFunction &IGF, CanType type,
+                                    MetadataResponse response) {
   auto key = LocalTypeDataKey(type, LocalTypeDataKind::forFormalTypeMetadata());
-  setScopedLocalTypeData(key, response);
+  IGF.setScopedLocalTypeData(key, response);
+}
+
+/// Walks the types upon whose corresponding metadata records' completeness the
+/// completeness of \p rootTy's metadata record depends.  For each such type,
+/// marks the corresponding locally cached metadata record, if any, complete.
+class TransitiveMetadataCompletion {
+  IRGenFunction &IGF;
+  CanType rootTy;
+  GraphNodeWorklist<CanType, 4> worklist;
+
+public:
+  TransitiveMetadataCompletion(IRGenFunction &IGF, CanType rootTy)
+      : IGF(IGF), rootTy(rootTy) {}
+
+  void complete();
+
+private:
+  /// Marks the metadata record currently locally cached corresponding to \p ty
+  /// complete.
+  ///
+  /// Returns whether \p ty's transitive metadata should be marked complete.
+  bool visit(CanType ty) {
+    // If it's the root type, it's already been marked complete, but we want to
+    // mark its transitively dependent metadata as complete.
+    if (ty == rootTy)
+      return true;
+    auto key = LocalTypeDataKey(ty, LocalTypeDataKind::forFormalTypeMetadata());
+    auto data = IGF.tryGetLocalTypeMetadata(key, MetadataState::Abstract);
```

Comment:
Hmm.  I think this is going to force abstract entries in the LocalTypeData cache to emitted (abstract in the sense of the LTDC, i.e. metadata that can cheaply be produced).  We probably don't want that; it might have a lot of extra compile-time cost, even if LLVM can eliminate it later.

It might be easiest to just add a new operation on the LTDC, some sort of update-state-in-scope operation.  The initial implementation can just affect concrete entries, but in principle we could also make this work with abstract entries by cloning the abstract path and then telling it to override the completeness.

Reply:
Fixed.  Added an `advanceStateInScope` operation to LTDC.  It looks for relevant cache entries and updates the cache accordingly.  There are TODOs for the abstract case.  If a relevant concrete entry is found (i.e. one whose dominance point dominates the active dominance point), it's used to produce a new cache entry whose dominance point is the active dominance point and whose state is the one specified.